docs(network): README (#2100)
ChainSafe/gossamer: dot/network/README.md

# Gossamer `network` Package

> This package emulates the [peer-to-peer networking capabilities](https://crates.parity.io/sc_network/index.html)

Author: Both? 😅 I think what I'm trying to say in this case, though, is that it "emulates" (seeks to mimic) them.

> This package emulates the [peer-to-peer networking capabilities](https://crates.parity.io/sc_network/index.html)

Reviewer: I'd probably say implements haha

> This package emulates the [peer-to-peer networking capabilities](https://crates.parity.io/sc_network/index.html)

Author: Done!

authority and privilege between the maintainers of the network and its users - in a P2P network, each participant
possesses equal authority and equal privilege. `libp2p` is a framework for implementing P2P networks that was
modularized out of [IPFS](https://ipfs.io/); there are implementations in many languages including Go (used by this
project), Rust, Javascript, C++, and more. In addition to the standard library of protocols in a `libp2p`
implementation, there is a rich ecosystem of P2P networking packages that work with the pluggable architecture of
`libp2p`. In some cases, Gossamer uses the `libp2p` networking primitives to implement custom protocols for
blockchain-specific use cases. What follows is an exploration into three concepts that underpin P2P networks: identity &
key management, peer discovery & management, and multiplexing.

### Identity & Key Management

Many peer-to-peer networks, including those built with Gossamer, use
[public-key cryptography](https://en.wikipedia.org/wiki/Public-key_cryptography) (also known as asymmetric cryptography)
to allow network participants to securely identify themselves and interact with one another. The term "asymmetric"

> refers to the fact that in a public-key cryptography system, each participant's identity is associated with a set of two

Reviewer 2: Maybe contrast asymmetric encryption where 2 keys are needed to symmetric encryption where there is only one key to highlight the difference between the two kinds of encryption

Author: I think I would rather keep the discussion tightly focused on asymmetric encryption as opposed to introducing new concepts that don't actually pertain to Gossamer.

keys, each of which serve a distinct ("asymmetric") purpose. One of the keys in an asymmetric key pair is public, this
is the key that the participant uses to identify themselves; the other pair is private and is used by the network
participant to "sign" messages in order to cryptographically prove that the message originated from the private key's
owner. It may be constructive to think about a public key as a username and private key as a password, such as for a
banking or social media website. Participants in P2P networks that use asymmetric cryptography must protect their
private keys, as well as maintain indices of the public keys that belong to the other participants in the network.

> Gossamer provides a [keystore](../../lib/keystore) for securely storing private keys. There are a number of Gossamer

Author: That is the purpose that the keystore serves...it allows for the secure management of the private keys that correspond with the node's public keys (in order to allow the node to sign messages and such).

> Gossamer provides a [keystore](../../lib/keystore) for securely storing private keys. There are a number of Gossamer

Reviewer 2: ohhh okay I misunderstood I thought you were saying we store other nodes private keys, makes sense

> Gossamer provides a [keystore](../../lib/keystore) for securely storing private keys. There are a number of Gossamer

Author: Okay, I actually amended the text slightly to clarify this ✅

processes that manage the public keys of network peers - some of these, such as
[peer discovery and management](#peer-discovery--management), are described in this document, but there are other
packages (most notably [`peerset`](../peerset)) that also interact with the public keys of network peers. One of the
most critical details in a network that uses asymmetric cryptography is the
[key distribution](https://en.wikipedia.org/wiki/Key_distribution) mechanism, which is the process that the hosts in the
network use to securely exchange public keys - `libp2p` supports [Noise](#noise), a key distribution framework that is

> based on [Diffie-Hellman key exchange](https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange).

Reviewer 2: Nice!

## Peer-to-Peer Networking & `libp2p`

[Peer-to-peer](https://en.wikipedia.org/wiki/Peer-to-peer) networking has been a dynamic field of research for over two
decades, and P2P protocols are at the heart of many blockchain networks. P2P networks can be contrasted with traditional
[client-server](https://en.wikipedia.org/wiki/Client%E2%80%93server_model) networks where there is a clear separation of
authority and privilege between the maintainers of the network and its users - in a P2P network, each participant
possesses equal authority and equal privilege. `libp2p` is a framework for implementing P2P networks that was
modularized out of [IPFS](https://ipfs.io/); there are implementations in many languages including Go (used by this
project), Rust, Javascript, C++, and more. In addition to the standard library of protocols in a `libp2p`
implementation, there is a rich ecosystem of P2P networking packages that work with the pluggable architecture of
`libp2p`. In some cases, Gossamer uses the `libp2p` networking primitives to implement custom protocols for
blockchain-specific use cases. What follows is an exploration into three concepts that underpin P2P networks:

> [identity & key management](<(#identities--key-management)>),

Reviewer 3: ```suggestion
[identity & key management](#identities--key-management),
```

# Gossamer `network` Package

> This package implements the [peer-to-peer networking capabilities](https://crates.parity.io/sc_network/index.html)
> provided by the [Substrate](https://docs.substrate.io/) framework for blockchain development, which implies that it is
> built on the extensible [`libp2p` networking stack](https://docs.libp2p.io/introduction/what-is-libp2p/). `libp2p`
> provides implementations of a number of battle-tested peer-to-peer (P2P) networking protocols (e.g. [Noise](#noise) for

Reviewer: ```suggestion
This package implements the [peer-to-peer networking capabilities](https://crates.parity.io/sc_network/index.html)
provided by the [Substrate](https://docs.substrate.io/) framework for blockchain development. It is
built on the extensible [`libp2p` networking stack](https://docs.libp2p.io/introduction/what-is-libp2p/). `libp2p`
provides implementations of a number of battle-tested peer-to-peer (P2P) networking protocols (e.g. [Noise](#noise) for
```

[key exchange](#identities--key-management), and [Yamux](#yamux) for [stream multiplexing](#stream-multiplexing)), and
also makes it possible to implement the blockchain-specific protocols defined by Substrate (e.g. [syncing](#sync) and
[finalising](#GRANDPA) blocks, and maintaining the [transaction pool](#transactions)). The purpose of this document is
to provide the information that is needed to understand the P2P networking capabilities that are implemented by
Gossamer - this includes an introduction to P2P networks and `libp2p`, as well as detailed descriptions of the Gossamer
P2P networking protocols.

## Peer-to-Peer Networking & `libp2p`

[Peer-to-peer](https://en.wikipedia.org/wiki/Peer-to-peer) networking has been a dynamic field of research for over two

> decades, and P2P protocols are at the heart of many blockchain networks. P2P networks can be contrasted with traditional

Reviewer: ```suggestion
decades, and P2P protocols are at the heart of blockchain networks. P2P networks can be contrasted with traditional
```

i hope all blockchains use p2p haha, maybe this isn'true? would be concerning if not

blockchain-specific use cases. What follows is an exploration into three concepts that underpin P2P networks:
[identity & key management](#identity--key-management), [peer discovery & management](#peer-discovery--management), and
[stream multiplexing](#stream-multiplexing).

### Identity & Key Management

Many peer-to-peer networks, including those built with Gossamer, use
[public-key cryptography](https://en.wikipedia.org/wiki/Public-key_cryptography) (also known as asymmetric cryptography)
to allow network participants to securely identify themselves and interact with one another. The term "asymmetric"
refers to the fact that in a public-key cryptography system, each participant's identity is associated with a set of two
keys, each of which serve a distinct ("asymmetric") purpose. One of the keys in an asymmetric key pair is private and is
used by the network participant to "sign" messages in order to cryptographically prove that the message originated from
the private key's owner; the other key is public, this is the key that the participant uses to identify themselves - it

> is distributed to network peers to allow for the decryption of messages signed by the corresponding public key. It may

Reviewer: ```suggestion
is distributed to network peers to allow for the verification of messages signed by the corresponding private key. It may
```

this was a bit confusing, it first talks about signing then about decryption?

be constructive to think about a public key as a username and private key as a password, such as for a banking or social

> media website. Participants in P2P networks that use asymmetric cryptography must protect their private keys, as well as
> maintain indices of the public keys that belong to the other participants in the network. Gossamer provides a

Reviewer: ```suggestion
media website. Participants in P2P networks that use asymmetric cryptography must protect their private keys, as well as
keep track of the public keys that belong to the other participants in the network. Gossamer provides a
```

exists in the TCP/IP stack, where unique port numbers are used to distinguish logically independent streams that share a
common physical transport medium. Gossamer uses [Yamux](#yamux) for stream multiplexing.

## Gossamer Network Protocols

### `libp2p`-Integrated Protocols

#### `ping`

This is a simple liveness check [protocol](https://docs.libp2p.io/concepts/protocols/#ping) that peers can use to
quickly see if another peer is online - it is
[included](https://github.com/libp2p/go-libp2p/tree/master/p2p/protocol/ping) with the official Go implementation of
`libp2p`.

#### `identify`

The [`identify` protocol](https://docs.libp2p.io/concepts/protocols/#identify) allows peers to exchange information
about each other, most notably their public keys and known network addresses; like [`ping`](#ping), it is
[included with `go-libp2p`](https://github.com/libp2p/go-libp2p/tree/master/p2p/protocol/identify).

#### Noise

As described above, [Noise](http://noiseprotocol.org/) provides `libp2p` with its
[key distribution](#identity--key-management) capabilities. The Noise protocol is

> [well documented](http://cryptowiki.net/index.php?title=Noise_Protocol_Framework) and the Go implementation is

Reviewer: it might be nicer to link to the libp2p spec: https://github.com/libp2p/specs/blob/master/noise/README.md afaik the libp2p version of noise and the "official" version have some slight differences

maintained [under the official](https://github.com/libp2p/go-libp2p-noise) `libp2p` GitHub organization. Noise defines a
number of [variables](http://cryptowiki.net/index.php?title=Noise_Protocol_Framework#Noise_Variables) and
[handshake patterns](http://cryptowiki.net/index.php?title=Noise_Protocol_Framework#Handshake_patterns) that
participants in a peer-to-peer network can use to establish message-passing channels with one another.

#### Yamux

[Yamux (Yet another Multiplexer)](https://github.com/hashicorp/yamux) is a Golang library for
[stream-oriented multiplexing](#stream-multiplexing) that is maintained by [HashiCorp](https://www.hashicorp.com/) - it
implements a well defined [specification](https://github.com/hashicorp/yamux/blob/master/spec.md). Gossamer uses
[the official `libp2p` adapter](https://github.com/libp2p/go-libp2p-yamux) for Yamux.

#### Kademlia

[Kademlia](https://en.wikipedia.org/wiki/Kademlia) is a battle-tested
[distributed hash table (DHT)](https://en.wikipedia.org/wiki/Distributed_hash_table) that defines methods for managing a

> dynamic list of peers that is constantly updated in order to make a P2P network more resilient and resistant to attacks.

Reviewer: I'd clarify this by explicitly saying the DHT is used for peer discovery, and that each node will advertise themselves in the DHT, and to discover peers nodes will "walk" the DHT.

Kademlia calculates a logical "distance" between any two nodes in the network by applying the xor operation to the IDs
of those two peers. Although this "distance" is not correlated to the physical distance between the peers, it adheres to
three properties that are [crucial to the analysis](https://en.wikipedia.org/wiki/Kademlia#Academic_significance) of
Kademlia as a protocol - in particular, these three properties are:

- the "distance" between a peer and itself is zero
- the "distance" between two peers is the same regardless of the order in which the peers are considered (it is
  [symmetric](https://en.wikipedia.org/wiki/Symmetry_in_mathematics))
- the shortest "distance" between two peers does not include any intermediate peers (it follows the

>   [triangle inequality](https://en.wikipedia.org/wiki/Triangle_inequality))

Reviewer 4: It might be worth mentioning that in Kademlia, peers save more information about peers that are closer to them. And because of this structure in peers, we can walk in the dht to the next closest peer to the peer that we are looking for.

There is a nice talk by feross, if you want to include such things https://youtu.be/kxHRATfvnlw?t=1029

Author: This is a little more detail than I'd like to get into in this document, but very interesting!

Gossamer uses [the official `libp2p` implementation of Kademlia for Go](https://github.com/libp2p/go-libp2p-kad-dht).

### Blockchain-Specific Protocols

The `libp2p` stack is used to implement the blockchain-specific protocols that are used to participate in
"Substrate-like" networks - these protocols are divided into two types, [notification](#notification-protocols) and
[request/response](#requestresponse-protocols). The two types of protocols are described in greater details below, along
with the specific protocols for each type.

##### Notification Protocols

[Notification protocols](https://crates.parity.io/sc_network/index.html#notifications-protocols) allow peers to
unidirectionally "push" information to other peers in the network. Although the peer receiving the information must
explicitly accept a handshake in order to open a stream for a notification protocol, this stream does not allow the

> receiver to "push" information to the sender.

Reviewer: I'd also update this to say something like "the stream can be left open indefinitely until one side no longer wishes to communicate"

###### Transactions

This protocol is used to notify network peers of [transactions](https://docs.substrate.io/v3/concepts/tx-pool/) that
have been locally received and validated. Transactions are used to access the
[public APIs of blockchain runtimes](https://docs.substrate.io/v3/concepts/extrinsics/#signed-transactions).

###### Block Announces

The block announce protocol is used to notify network peers of the existence of a block. The message for this protocol
contains a [block header](https://docs.substrate.io/v3/getting-started/glossary/#header) and associated data, such as

> the [BABE pre-runtime digest](https://crates.parity.io/sp_consensus_babe/digests/enum.PreDigest.html).

Reviewer: the BABE digests are part of the header, so I'd just leave this as saying the message is a block header.

###### GRANDPA

[Finality](https://wiki.polkadot.network/docs/learn-consensus#finality-gadget-grandpa) protocols ("gadgets") such as
GRANDPA are often described in terms of "games" that are played by the participants in a network. In GRANDPA, this game

> relates to voting on the blocks that inform the canonical chain. This notification protocol is used by peers to cast

Reviewer: ```suggestion
relates to voting on what blocks shoudl be part of the canonical chain. This notification protocol is used by peers to cast
```

> votes for participation in the GRANDPA game.

Reviewer: ```suggestion
votes for participation in the GRANDPA game, and as well to gossip a `Commit` message when a chain has been finalised.
```

##### Request/Response Protocols

[These protocols](https://crates.parity.io/sc_network/index.html#request-response-protocols) allow peers to request
specific information from one another. The requesting peer sends a protocol-specific message that describes the request

> and the peer to which the request was sent replies with a message.

Reviewer: I'd update this to also say if a node opens a stream with some remote peer, then the outgoing side will *only* push requests and the incoming side will only expect to receive requests. same for the other way around, the outgoing side expects to only receive responses.
